src/api.rs

```diff
@@ -928,6 +985,75 @@ impl Client {
         }
     }
 
+    async fn obtain_sso_code(
+        &self,
+        sso_id: &str,
+    ) -> Result<(String, String, String)> {
+        let state: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(64)
+            .map(char::from)
+            .collect();
+
+        let sso_code_verifier: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(64)
+            .map(char::from)
+            .collect();
+
+        let mut hasher = Sha256::new();
+        hasher.update(sso_code_verifier.clone());
+        let code_challenge =
+            crate::base64::encode_url_safe_no_pad(hasher.finalize());
+
+        let Some(port) = (8065..8070)
+            .find(|&port| TcpListener::bind(("127.0.0.1", port)).is_ok())
+        else {
+            return Err(Error::FailedToFindFreePort {
+                range: "(8065..8070)".to_string(),
+            });
+        };
+
+        let listener = match TcpListener::bind(("127.0.0.1", port)) {
+            Ok(value) => value,
+            Err(e) => return Err(Error::CreateSSOCallbackServer { err: e }),
+        };
+
+        let callback_server =
+            async { start_sso_callback_server(&listener, state.as_str()) };
+
+        let callback_url =
+            "http://localhost:".to_string() + port.to_string().as_str();
+
+        if let Err(e) = open::that(
+            self.ui_url.clone()
+                + "/#/sso?clientId="
+                + "cli"
+                + "&redirectUri="
+                + urlencoding::encode(callback_url.as_str())
+                    .into_owned()
+                    .as_str()
+                + "&state="
+                + state.as_str()
+                + "&codeChallenge="
+                + code_challenge.as_str()
+                + "&identifier="
+                + sso_id,
+        ) {
+            // TODO: probably it'd be better to display the URL in the console if the automatic
+            // open operation fails, instead of failing the whole process? E.g. docker container
+            // case
```

Comment:
yeah, i'd go farther here and say that we should explicitly support both options here (via a command line argument to `rbw login` or similar)